StRoot/StSpinPool/StTriggerFilterMaker/StTriggerFilterMaker.cxx

```diff
@@ -21,39 +22,85 @@ Int_t StTriggerFilterMaker::Init() {
+    StTriggerIdCollection* trgIdColl = event->triggerIdCollection();
+    const StTriggerData* trgdata = event->triggerData();
     if(trgIdColl){
-	if(mPrint){
-	    vector<unsigned int> ids = trgIdColl->nominal()->triggerIds();
-	    LOG_INFO << "Offline Trigger Id = ";
-	    for(unsigned i=0; i<ids.size(); ++i) LOG_INFO << ids[i] << " ";
-	}
-	for(unsigned i=0; i<mVetoTriggers.size(); ++i) {
-	    if(trgIdColl->nominal()->isTrigger(mVetoTriggers[i])) {
-		if(mPrint)  LOG_INFO << " Veto" << endm;
-		return kStSkip;
-	    }
+      if(mPrint){
+	vector<unsigned int> ids = trgIdColl->nominal()->triggerIds();
+	LOG_INFO << "Offline Trigger Id = ";
+	for(unsigned i=0; i<ids.size(); ++i) LOG_INFO << ids[i] << " ";
+      }
+      for(unsigned i=0; i<mVetoTriggers.size(); ++i) {
+	if(trgIdColl->nominal()->isTrigger(mVetoTriggers[i])) {
+	  if(mPrint)  LOG_INFO << " Vetoed" << endm;
+	  return kStSkip;
 	}
-	for(unsigned i=0; i<mGoodTriggers.size(); ++i) {
-	    if(trgIdColl->nominal()->isTrigger(mGoodTriggers[i])) {
-		if(mPrint)   LOG_INFO << " Accept" << endm;
-		return kStOk;
+      }
+      for(unsigned i=0; i<mGoodTriggers.size(); ++i) {
+	if(trgIdColl->nominal()->isTrigger(mGoodTriggers[i])) {
+	  
+	  if(trgdata && mTofUpperLimit>0){
+	    unsigned int tof = trgdata->tofMultiplicity();
+	    if(tof < mTofUpperLimit){
+	      if(mPrint) LOG_INFO << " Accepted with TOF upper limit" << endm;
+	      return kStOk;
+	    }else{
+	      if(mPrint) 
+		LOG_INFO 
+		  << Form(" TriggerId accepted but rejected with TOFmult uppter limit TOFM=%d < UpperLimit=%d",
+			  tof,mTofUpperLimit)
+		  << endm;
+	      return kStSkip;
 	    }
+	  }
+	  
+	  if(mPrint) LOG_INFO << " Accepted (no TOF condition)" << endm;
+	  return kStOk;
 	}
-
+      }
+    }
+  }else if(StMuDst::event()){    
+    StMuTriggerIdCollection& trgIdColl = StMuDst::event()->triggerIdCollection();
+    const StTriggerData* trgdata = StMuDst::event()->triggerData();
+    if(mPrint){
+      vector<unsigned int> ids = trgIdColl.nominal().triggerIds();
+      LOG_INFO << "Offline Trigger Id = ";
+      for(unsigned i=0; i<ids.size(); ++i) LOG_INFO << ids[i] << " ";
+    }
+    for(unsigned i=0; i<mVetoTriggers.size(); ++i) {
+      if(trgIdColl.nominal().isTrigger(mVetoTriggers[i])) {
+	if(mPrint)  LOG_INFO << " Vetoed" << endm;
+	return kStSkip;
+      }
+    }
+    for(unsigned i=0; i<mGoodTriggers.size(); ++i) {
+      if(trgIdColl.nominal().isTrigger(mGoodTriggers[i])) {
+	
+	if(trgdata && mTofUpperLimit>0){
+	  unsigned int tof = trgdata->tofMultiplicity();
+	  if(tof < mTofUpperLimit){
+	    if(mPrint) LOG_INFO << " Accepted with TOF upper limit" << endm;
+	    return kStOk;
+	  }else{
+	    if(mPrint) 
+	      LOG_INFO 
+		<< Form(" TriggerId accepted but rejected with TOFmult uppter limit TOFM=%d < UpperLimit=%d",
+			tof,mTofUpperLimit)
+		<< endm;
+	    return kStSkip;
+	  }
+	}	  
+	if(mPrint) LOG_INFO << " Accepted (no TOF condition)" << endm;
+	return kStOk;
+      }
```

Comment:
This looks like an exact copy-paste of the StEvent case above, except you've seemingly lost the `if(trgIdColl){`. Could we de-duplicate this?

Reply:
Yes, except trgIdColl is StTriggerIdCollection* in StEvent case and StMuTriggerIdCollection& in MuDst case.  Their nominal() returns const StTriggerId* and 
const StTriggerId&. I tried to make it one piece of code but I could not come up with a good compromise. I cannot out if(trgIdColl) when it is a reference not a pointer, no?

Reply:
> Their nominal() returns const StTriggerId* and const StTriggerId&.

That part is non-issue, you can always use std::remove_pointer https://www.cplusplus.com/reference/type_traits/remove_pointer/

The difference between St and StMu classes is a bit tricky. Maybe we could go with something like:

```diff
diff --git a/StRoot/StSpinPool/StTriggerFilterMaker/StTriggerFilterMaker.cxx b/StRoot/StSpinPool/StTriggerFilterMaker/StTriggerFilterMaker.cxx
index 16ef1a3ce1..42f2b41ef0 100644
--- a/StRoot/StSpinPool/StTriggerFilterMaker/StTriggerFilterMaker.cxx
+++ b/StRoot/StSpinPool/StTriggerFilterMaker/StTriggerFilterMaker.cxx
@@ -1,5 +1,7 @@
 // $Id: StTriggerFilterMaker.cxx,v 1.3 2021/04/19 18:49:54 akio Exp $
 
+#include <functional>
+
 #include "StTriggerFilterMaker.h"
 #include "StMuDSTMaker/COMMON/StMuDst.h"
 #include "StMuDSTMaker/COMMON/StMuEvent.h"
@@ -22,85 +24,61 @@ Int_t StTriggerFilterMaker::Init() {
 }
 
 Int_t StTriggerFilterMaker::Make() {
-  StEvent* event= (StEvent*)GetInputDS("StEvent");
-  if(event){
-    StTriggerIdCollection* trgIdColl = event->triggerIdCollection();
-    const StTriggerData* trgdata = event->triggerData();
-    if(trgIdColl){
-      if(mPrint){
-	vector<unsigned int> ids = trgIdColl->nominal()->triggerIds();
-	LOG_INFO << "Offline Trigger Id = ";
-	for(unsigned i=0; i<ids.size(); ++i) LOG_INFO << ids[i] << " ";
-      }
-      for(unsigned i=0; i<mVetoTriggers.size(); ++i) {
-	if(trgIdColl->nominal()->isTrigger(mVetoTriggers[i])) {
-	  if(mPrint)  LOG_INFO << " Vetoed" << endm;
-	  return kStSkip;
-	}
-      }
-      for(unsigned i=0; i<mGoodTriggers.size(); ++i) {
-	if(trgIdColl->nominal()->isTrigger(mGoodTriggers[i])) {
-	  
-	  if(trgdata && mTofUpperLimit>0){
-	    unsigned int tof = trgdata->tofMultiplicity();
-	    if(tof < mTofUpperLimit){
-	      if(mPrint) LOG_INFO << " Accepted with TOF upper limit" << endm;
-	      return kStOk;
-	    }else{
-	      if(mPrint) 
-		LOG_INFO 
-		  << Form(" TriggerId accepted but rejected with TOFmult uppter limit TOFM=%d < UpperLimit=%d",
-			  tof,mTofUpperLimit)
-		  << endm;
-	      return kStSkip;
-	    }
-	  }
-	  
-	  if(mPrint) LOG_INFO << " Accepted (no TOF condition)" << endm;
-	  return kStOk;
-	}
-      }
-    }
-  }else if(StMuDst::event()){    
-    StMuTriggerIdCollection& trgIdColl = StMuDst::event()->triggerIdCollection();
-    const StTriggerData* trgdata = StMuDst::event()->triggerData();
-    if(mPrint){
-      vector<unsigned int> ids = trgIdColl.nominal().triggerIds();
-      LOG_INFO << "Offline Trigger Id = ";
-      for(unsigned i=0; i<ids.size(); ++i) LOG_INFO << ids[i] << " ";
+    StEvent* event= (StEvent*)GetInputDS("StEvent");
+    const StTriggerData* trgdata;
+    vector<unsigned int> triggerIds;
+    std::function<bool(unsigned int)> isTrigger;
+    if(event) {
+       StTriggerIdCollection* trgIdColl = event->triggerIdCollection();
+       trgdata = event->triggerData();
+       triggerIds = trgIdColl->nominal()->triggerIds();
+       isTrigger = std::bind(&StTriggerId::isTrigger, trgIdColl->nominal(), std::placeholders::_1);
+    } else if(StMuDst::event()) {
+       StMuTriggerIdCollection& trgIdColl = StMuDst::event()->triggerIdCollection();
+       triggerIds = trgIdColl.nominal().triggerIds();
+       trgdata = StMuDst::event()->triggerData();
+       isTrigger = std::bind(&StTriggerId::isTrigger, &trgIdColl.nominal(), std::placeholders::_1);
+    } else {
+       LOG_INFO << "No StEvent nor Mudst" << endm;
     }
-    for(unsigned i=0; i<mVetoTriggers.size(); ++i) {
-      if(trgIdColl.nominal().isTrigger(mVetoTriggers[i])) {
-	if(mPrint)  LOG_INFO << " Vetoed" << endm;
-	return kStSkip;
-      }
-    }
-    for(unsigned i=0; i<mGoodTriggers.size(); ++i) {
-      if(trgIdColl.nominal().isTrigger(mGoodTriggers[i])) {
-	
-	if(trgdata && mTofUpperLimit>0){
-	  unsigned int tof = trgdata->tofMultiplicity();
-	  if(tof < mTofUpperLimit){
-	    if(mPrint) LOG_INFO << " Accepted with TOF upper limit" << endm;
-	    return kStOk;
-	  }else{
-	    if(mPrint) 
-	      LOG_INFO 
-		<< Form(" TriggerId accepted but rejected with TOFmult uppter limit TOFM=%d < UpperLimit=%d",
-			tof,mTofUpperLimit)
-		<< endm;
-	    return kStSkip;
-	  }
-	}	  
-	if(mPrint) LOG_INFO << " Accepted (no TOF condition)" << endm;
-	return kStOk;
-      }
+
+    if (trgdata) {
+       if (mPrint) {
+          LOG_INFO << "Offline Trigger Id = ";
+          for(unsigned int id : triggerIds) LOG_INFO << id << " ";
+       }
+       for(unsigned int id_veto : mVetoTriggers) {
+          if (isTrigger(id_veto)) {
+             if (mPrint)  LOG_INFO << " Vetoed" << endm;
+             return kStSkip;
+          }
+       }
+       for(unsigned int id_good : mGoodTriggers) {
+          if (isTrigger(id_good)) {
+
+             if (mTofUpperLimit>0) {
+                unsigned int tof = trgdata->tofMultiplicity();
+                if (tof < mTofUpperLimit) {
+                   if (mPrint) LOG_INFO << " Accepted with TOF upper limit" << endm;
+                   return kStOk;
+                } else {
+                   if (mPrint)
+                      LOG_INFO
+                         << Form(" TriggerId accepted but rejected with TOFmult uppter limit TOFM=%d < UpperLimit=%d",
+                                 tof,mTofUpperLimit)
+                         << endm;
+                   return kStSkip;
+                }
+             }
+
+             if (mPrint) LOG_INFO << " Accepted (no TOF condition)" << endm;
+             return kStOk;
+          }
+       }
     }
-  }else{
-    LOG_INFO << "No StEvent nor Mudst" << endm;
-  }	
-  if(mPrint) LOG_INFO << "Skip" << endm;
-  return kStSkip;
+
+    if(mPrint) LOG_INFO << "Skip" << endm;
+    return kStSkip;
 }
 
 /*****************************************************************************
```